# Module 4 Assignment

### Gabi Rivera    ||   20Nov2022   ||   ADS502-01

In [1]:
import os
os.getcwd()

'/Users/gabirivera/Desktop/MSADS2/ADS502-01/Module4/Assignment'

In [25]:
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier

### Data Science Using Python and R: Chapter 13

13. Create a logistic regression model to predict whether or not a customer has a store credit card, based on whether they have a web account and the days between purchases. Obtain the summary of the model.

Import train and test datasets: 

In [7]:
cs_train = pd.read_csv("clothing_sales_training.csv", sep = ",")
cs_train.head()

CC   Days  Web  Sales per Visit
0   0  333.0    0       184.230000
1   0  171.5    0        38.500000
2   0  213.0    0       150.326667
3   1   71.4    1       104.240000
4   1  145.0    0       782.080000

In [6]:
cs_test = pd.read_csv("clothing_sales_test.csv", sep = ",")
cs_test.head()

CC    Days  Web  Sales per Visit
0   1  174.00    0          64.5000
1   1   87.62    0         105.7575
2   0   49.00    0          87.4400
3   0   72.50    0          60.0000
4   0  264.00    0         318.5000

Prepare X and y variables:

In [9]:
X = pd.DataFrame(cs_train[['Days', 'Web']])
X = sm.add_constant(X)

y = pd.DataFrame(cs_train[['CC']])

Create logistic regression: 

In [12]:
logreg01 = sm.Logit(y, X).fit()
logreg01.summary2()

Optimization terminated successfully.
         Current function value: 0.655955
         Iterations 5


<class 'statsmodels.iolib.summary2.Summary'>
"""
                         Results: Logit
=================================================================
Model:              Logit            Pseudo R-squared: 0.053     
Dependent Variable: CC               AIC:              1909.5825 
Date:               2022-11-20 10:00 BIC:              1925.4226 
No. Observations:   1451             Log-Likelihood:   -951.79   
Df Model:           2                LL-Null:          -1004.9   
Df Residuals:       1448             LLR p-value:      8.3668e-24
Converged:          1.0000           Scale:            1.0000    
No. Iterations:     5.0000                                       
-------------------------------------------------------------------
           Coef.    Std.Err.      z      P>|z|     [0.025    0.975]
-------------------------------------------------------------------
const      0.4962     0.0887    5.5968   0.0000    0.3224    0.6699
Days      -0.0037     0.0004   -8.4491   0.0000   -0.0046   -0.0028
Web        1.2537     0.3307    3.7914   0.0001    0.6056    1.9018
=================================================================

"""

14. Are there any variables that should be removed from the model? If so, remove them and rerun the model.

Both variables will be included because there's no sign of multicollinearity amongst the variables. The z-values have reasonable scores with significantly small p-values less than 0.05 level of significance.

15. Write the descriptive form of the logistic regression model using the coefficients obtained from Question 1.

ln(CC) = 0.4962 -0.0037Days + 1.2537Web

16. Validate the model using the test data set.

In [13]:
X_test = pd.DataFrame(cs_test[['Days', 'Web']])
X_test = sm.add_constant(X_test)
y_test = pd.DataFrame(cs_test[['CC']])

logreg01_test = sm.Logit(y_test, X_test).fit()
logreg01_test.summary2()

Optimization terminated successfully.
         Current function value: 0.656885
         Iterations 5


<class 'statsmodels.iolib.summary2.Summary'>
"""
                         Results: Logit
=================================================================
Model:              Logit            Pseudo R-squared: 0.052     
Dependent Variable: CC               AIC:              1838.7104 
Date:               2022-11-20 10:02 BIC:              1854.4324 
No. Observations:   1395             Log-Likelihood:   -916.36   
Df Model:           2                LL-Null:          -966.40   
Df Residuals:       1392             LLR p-value:      1.8534e-22
Converged:          1.0000           Scale:            1.0000    
No. Iterations:     5.0000                                       
-------------------------------------------------------------------
           Coef.    Std.Err.      z      P>|z|     [0.025    0.975]
-------------------------------------------------------------------
const      0.4634     0.0873    5.3105   0.0000    0.2924    0.6345
Days      -0.0035     0.0004   -8.2261   0.0000   -0.0043   -0.0026
Web        1.0973     0.2830    3.8780   0.0001    0.5427    1.6519
=================================================================

"""

17. Obtain the predicted values of the response variable for each record in the data set.

In [ ]:
Predicted values fo test datast: 

In [17]:
predictions_prob = logreg01.predict(X_test)
predictions_prob.head()

0    0.463090
1    0.542853
2    0.578054
3    0.556706
4    0.382003
dtype: float64

In [ ]:
Convert predictions to binary results: 

In [18]:
predictions = (logreg01.predict(X_test) > 0.5).astype(int)
predictions.head()

0    0
1    1
2    1
3    1
4    0
dtype: int64

### Data Science Using Python and R: Chapter 6

19. Use random forests on the training data set to predict income using marital status and capital gains and losses.

In [23]:
a_train = pd.read_csv("adult_ch6_training", sep = ",")
a_train.head()

Marital status Income  Cap_Gains_Losses
0  Never-married  <=50K           0.02174
1       Divorced  <=50K           0.00000
2        Married  <=50K           0.00000
3        Married  <=50K           0.00000
4        Married  <=50K           0.00000

In [24]:
a_test = pd.read_csv("adult_ch6_test", sep = ",")
a_test.head()

Marital status Income  Cap_Gains_Losses
0        Married  <=50K          0.000000
1        Married   >50K          0.051781
2  Never-married  <=50K          0.000000
3       Divorced   >50K          0.000000
4        Married   >50K          0.000000

Transform dependent variable to one-dimention: 

In [53]:
y = a_train[['Income']]
rfy = np.ravel(y)

mar_np = np.array(a_train['Marital status'])
(mar_cat, mar_cat_dict) = stattools.categorical(mar_np, drop=True, dictnames = True)

mar_cat_pd = pd.DataFrame(mar_cat)
X = pd.concat((a_train[['Cap_Gains_Losses']], mar_cat_pd), axis = 1)

/Users/gabirivera/opt/anaconda3/lib/python3.8/site-packages/statsmodels/tools/tools.py:152: FutureWarning: categorical is deprecated. Use pandas Categorical to represent categorical data and can get_dummies to construct dummy arrays. It will be removed after release 0.13.
  warnings.warn(


Create random forrets train model: 

In [54]:
rf01 = RandomForestClassifier(n_estimators = 100, criterion="gini").fit(X,rfy)

/Users/gabirivera/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:1858: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  warnings.warn(


In [56]:
rf01_d = rf01.predict(X)
a_train['pred'] = rf01_d.tolist()
a_train.head()

/Users/gabirivera/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:1858: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  warnings.warn(


Marital status Income  Cap_Gains_Losses   pred
0  Never-married  <=50K           0.02174  <=50K
1       Divorced  <=50K           0.00000  <=50K
2        Married  <=50K           0.00000  <=50K
3        Married  <=50K           0.00000  <=50K
4        Married  <=50K           0.00000  <=50K

In [60]:
a_train['pred'].value_counts()

<=50K    17375
>50K      1386
Name: pred, dtype: int64

In [68]:
(1386/(17375+1386)*100)

7.387665902670433

20. Use random forests using the test data set that utilizes the same target and predictor variables.Does the test data result match the training data result?

Transform test dataset's dependent variable to one-dimention: 

In [32]:
y2 = a_test[['Income']]
rfy2 = np.ravel(y2)

mar_np2 = np.array(a_test['Marital status'])
(mar_cat2, mar_cat_dict2) = stattools.categorical(mar_np2, drop=True, dictnames = True)

mar_cat_pd2 = pd.DataFrame(mar_cat2)
X2 = pd.concat((a_test[['Cap_Gains_Losses']], mar_cat_pd2), axis = 1)

/Users/gabirivera/opt/anaconda3/lib/python3.8/site-packages/statsmodels/tools/tools.py:152: FutureWarning: categorical is deprecated. Use pandas Categorical to represent categorical data and can get_dummies to construct dummy arrays. It will be removed after release 0.13.
  warnings.warn(


In [33]:
rf02 = RandomForestClassifier(n_estimators = 100, criterion="gini").fit(X2,rfy2)

/Users/gabirivera/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:1858: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  warnings.warn(


In [48]:
rf02_d = rf02.predict(X)
a_test['pred'] = rf02_d.tolist()

/Users/gabirivera/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:1858: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  warnings.warn(


In [49]:
a_test.head()

Marital status Income  Cap_Gains_Losses   pred
0        Married  <=50K          0.000000  <=50K
1        Married   >50K          0.051781   >50K
2  Never-married  <=50K          0.000000  <=50K
3       Divorced   >50K          0.000000  <=50K
4        Married   >50K          0.000000  <=50K

In [59]:
a_test['pred'].value_counts()

<=50K    5702
>50K      453
Name: pred, dtype: int64

In [67]:
(453/(5702+453)*100)

7.359870024370431